include/nbl/builtin/hlsl/subgroup/arithmetic_portability_impl.hlsl

```diff
@@ -0,0 +1,490 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_SUBGROUP_ARITHMETIC_PORTABILITY_IMPL_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SUBGROUP_ARITHMETIC_PORTABILITY_IMPL_INCLUDED_
+
+#include "nbl/builtin/hlsl/binops.hlsl"
+#include "nbl/builtin/hlsl/subgroup/scratch.hlsl"
+#include "nbl/builtin/hlsl/subgroup/shuffle_portability.hlsl"
+
+// REVIEW:  Location and need of these. They need to be over a function but
+//          there's no need to have them over every subgroup func.
+//          The compiler doesn't seem to whine about these missing
+//          for subgroup*Min/Max functions even though the spec seems
+//          to mandate them https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Group_Operation
+[[vk::ext_extension("GL_KHR_shader_subgroup_basic")]]
+[[vk::ext_extension("GL_KHR_shader_subgroup_arithmetic")]]
+[[vk::ext_capability(/* GroupNonUniformArithmetic */ 63)]]
+[[vk::ext_capability(/* GroupNonUniformBallot */ 64)]]
+void fake_for_capability_and_extension(){}
+
+//const uint LastWorkgroupInvocation = _NBL_HLSL_WORKGROUP_SIZE_-1; // REVIEW: Where should this be defined?
+#define LastWorkgroupInvocation (_NBL_HLSL_WORKGROUP_SIZE_-1U)
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace subgroup
+{
+
+#ifdef NBL_GL_KHR_shader_subgroup_arithmetic
+namespace native
+{
+
+template<typename T, class Binop>
+struct reduction;
+template<typename T, class Binop>
+struct exclusive_scan;
+template<typename T, class Binop>
+struct inclusive_scan;
+
+// *** AND ***
+template<typename T>
+struct reduction<T, binops::bitwise_and<T> >
+{
+    T operator()(const T x)
+    {
+        return WaveActiveBitAnd(x);
+    }
+};
```

Comment:
oooh, so partial template specializations are possible like that?

Reply:
btw according to HLSL docs, this will only work for integer types, so best you implement with SPIR-V intrinsics so you have access to `float`